src/docs/arch-hs.dox

```diff
@@ -4,5 +4,111 @@ The History Store in WiredTiger tracks old (all but the latest committed)
+newest update to a key is written to the user table while the older updates
+for the key are written to the History Store. All user tables in the database are
+backed up by a single History Store table. History Store has no outward
+visibility to the application code and only WiredTiger's internal modules can
+perform operations on the History Store table using a predefined cursor API.
+
+@section arch_hs_table Historty Store table structure
+WiredTiger uses a standard row-store table to implement History Store internally.
+The key is a combination of:
+- \c btree ID of user table this update belongs to
+- \c record key (byte-string for row-store, record number for column-store)
+- \c start timestamp for the update
+- \c counter value
+This key format above allows us to search efficiently for a given record key and read
+timestamp combination. The last part of the key is a monotonically increasing counter
+to keep the key unique in scenarios where we have multiple transactions committing at
+the same timestamp. As the key for the History Store table is different for row- and column-store,
+we store both keytypes in a WT_ITEM because otherwise we'd need two history store files with
+different key formats. We can make the History Store table's key standard by moving the source
+key into the History Store table value, but that doesn't make the coding any simpler, and it makes
+the History Store table's value more likely to overflow the page size when the row-store key is
+relatively large.
+
+The corresponding value for each key is stored as a combination of:
+- \c stop timestamp
+- \c durable timestamp
+- \c update type
+- \c value.
+
+Note that every update in the Hisotry Store has a valid stop timestamp and transaction associated
+with it, except for cases where the update preceded a prepared update. The stop timestamp
+is fixed in this case once prepared update is resolved.
+
+@section arch_hs_initialize Historty Store initialization
+WiredTiger checks for an existing History Store table at startup and creates a new table if it is
+not able to find one. History Store table is usually created when WiredTiger creates a new database,
+with one exception where WiredTiger is openning an existing database created by an older version of
+WiredTiger that didn't support Histoty Store feature. WiredTiger uses \c WiredTigerHS.wt as the
+filename for the History Stable table.
+
+@section arch_hs_cursor Historty Store cursor interface
+WiredTiger uses a cursor interface for History Store table to make it easier for different modules
+to perform data operations on the History Strore table. A new cursor can be opened by calling
+\c __wt_curhs_open(). History Store cursor implementation supports most of the standard \c WT_CURSOR
+methods, with one key exception being \c WT_CURSOR::search() method. Recall that the key contains a
+counter value and it is expected that caller can not possibly know the correct counter value to use.
```

Comment:
I think the main reason is not because of counter but timestamp. We don't know whether the exact timestamp exists in hs or not. The search near fundamentally is to locate the cursor to the key with the closest timestamp.